ios/Exponent/Versioned/Core/UniversalModules/EXScopedErrorRecoveryModule.m

```diff
@@ -0,0 +1,71 @@
+// Copyright 2018-present 650 Industries. All rights reserved.
+
+#if __has_include(<EXErrorRecovery/EXErrorRecoveryModule.h>)
+#import "EXScopedErrorRecoveryModule.h"
+
+@interface EXScopedErrorRecoveryModule ()
+
+@property (nonatomic, strong) NSString *experienceId;
+
+@end
+
+@implementation EXScopedErrorRecoveryModule
+
+- (instancetype)initWithExperienceId:(NSString *)experienceId
+{
+  if (self = [super init]) {
+    _experienceId = experienceId;
+  }
+  return self;
+}
+
+- (BOOL)setRecoveryProps:(NSString *)props
+{
+  NSUserDefaults *preferences = [NSUserDefaults standardUserDefaults];
+  NSDictionary *errorRecoveryStore = [preferences objectForKey:[self getUserDefaultsKey]];
+  if (!errorRecoveryStore) {
+    return [EXScopedErrorRecoveryModule updateUserDefaults:preferences
+                                    withErrorRecoveryStore:@{ _experienceId: props }
+                                                   withKey:[self getUserDefaultsKey]];
+  } else {
+    NSMutableDictionary *propsToSave = [errorRecoveryStore mutableCopy];
+    [propsToSave setObject:props forKey:_experienceId];
+    return [EXScopedErrorRecoveryModule updateUserDefaults:preferences
+                                    withErrorRecoveryStore:propsToSave
+                                                   withKey:[self getUserDefaultsKey]];
+  }
+}
+
+- (NSString *)consumeRecoveryProps
+{
+  NSUserDefaults *preferences = [NSUserDefaults standardUserDefaults];
+  NSDictionary *errorRecoveryStore = [preferences objectForKey:[self getUserDefaultsKey]];
+  if (errorRecoveryStore) {
+    NSString *props = [errorRecoveryStore objectForKey:_experienceId];
+    if (props) {
+      NSMutableDictionary *storeWithRemovedProps = [errorRecoveryStore mutableCopy];
+      [storeWithRemovedProps removeObjectForKey:_experienceId];
+      
+      [EXScopedErrorRecoveryModule updateUserDefaults:preferences
+                               withErrorRecoveryStore:storeWithRemovedProps
+                                              withKey:[self getUserDefaultsKey]];
+      return props;
+    }
+  }
+  return nil;
+}
+
++ (BOOL)updateUserDefaults:(NSUserDefaults *)preferences
+    withErrorRecoveryStore:(NSDictionary *)newStore
+                   withKey:(NSString *)key
+{
+  [preferences setObject:newStore forKey:key];
+  return [preferences synchronize];
+}
+
+- (NSString *)getUserDefaultsKey
+{
+  return [NSString stringWithFormat:@"EX.%@", [super getUserDefaultsKey]];
```

Comment:
Does this choice of naming mean anything? seems better to do something like scoped.expo.errorRecovery.

Reply:
We don't want to `recovery props` propagate from one version to another so I've added `EX` to user defaults key. Then our versioning script will change it to something like `ABI_X_X.expo.errorRecovery`.

Reply:
1. Add a comment explaining this
2. We need to scope this per app

Reply:
I've scoped this per application.
In the scoped module, users defaults contained a dictionary with experiences id as a key. Then to every experience id, another dictionary with recovery props is assigned.

Reply:
OK, I see, the key is not scoped because the dictionary is the scoping mechanism.